pkg/networkservice/common/connect/client.go

```diff
@@ -68,25 +64,14 @@ func (u *connectClient) init() error {
 		var cc *grpc.ClientConn
 		cc, u.dialErr = grpc.DialContext(dialCtx, grpcutils.URLToTarget(clientURL), dialOptions...)
 		if u.dialErr != nil {
-			cancel()
 			return
 		}
 
 		u.client = u.clientFactory(u.ctx, cc)
 
 		go func() {
-			defer func() {
-				cancel()
-				_ = cc.Close()
-			}()
-			for cc.WaitForStateChange(u.ctx, cc.GetState()) {
-				switch cc.GetState() {
-				case connectivity.Connecting, connectivity.Idle, connectivity.Ready:
-					continue
-				default:
-					return
-				}
-			}
+			<-u.ctx.Done()
```

Comment:
Keep in mind that if the cancel function for u.ctx is called, this won't proceed to close the connection until some point *after* that.  There are no guarantees that it happens before some action you take immediately after calling the cancel function.

This may (or may not) be a problem depending on interactions with other parts of the system.

Reply:
You are right and this actually could be a problem while `ctx.Error` was checked only on Request error.
Fixed code to first check `ctx.Error` and after make a Request - it doesn't affect anything except the heal but does guarantee that restore Request will always meet canceled or already restarting connect client.